homeassistant/components/sungrow/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""The Sungrow Solar Energy integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from urllib.parse import ParseResult, urlparse
+
+from SungrowModbusClient import Client, sungrow_register
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import update_coordinator
+
+from .config_flow import CannotConnect
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Sungrow Solar Energy from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = SungrowData(hass, entry)
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class SungrowData(update_coordinator.DataUpdateCoordinator):
+    """Get and update the latest data."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the data object."""
+        super().__init__(
+            hass, _LOGGER, name="Sungrow", update_interval=timedelta(seconds=300)
+        )
+
+        self.host = entry.data.get(CONF_HOST, "")
+        self.port = entry.data.get(CONF_PORT, 502)
+
+        self.unique_id = entry.entry_id
+        self.name = entry.title
+
+        url = urlparse(self.host, "http")
+        netloc = url.netloc or url.path
+        path = url.path if url.netloc else ""
+        url = ParseResult("http", netloc, path, *url[3:])
```

Comment:
I reduced the string manipulation to a simpler version. This was unnecessarily convoluted.

Reply:
the reduced string manipulation was not able to take into account having "http://1.1.1.1" or "1.1.1.1" as input. so i added comments

Reply:
I think that the comment should say something like "Allow user to enter URLs with or without 'http://' at the start", if that is the intention of the code.

Reply:
I also recommend a parametrised pytest for this so that you can verify that the different URL schemes you want to accept all still work.

But that might be best in a follow up PR.